client/charts/js/components/DataLoader.jsx

```diff
@@ -8,39 +8,87 @@ function Loader() {
 }
 
 function DataLoader({ dataUrl, dataParser, dataKey, loadingComponent, children }) {
-	const [data, setData] = useState([])
+	const [data, setData] = useState({})
 	const [loading, setLoading] = useState(true)
+	const [fetchCache, setFetchCache] = useState({})
 
 	// Every time dataUrl or dataParser changes, fetch new data and parse it
 	useEffect(() => {
-		(async () => {
+		(() => {
 			setLoading(true)
-			const response = await fetch(dataUrl)
-			const content = await response.text()
-			const parsed = dataParser(content)
-			setData(parsed)
-			setLoading(false)
+
+			// Convert singular strings into arrays
+			const dataUrls = (typeof dataUrl === "string" ? [dataUrl] : dataUrl).filter(d => d)
+			const dataKeys = (typeof dataKey === "string" ? [dataKey] : dataKey).filter(d => d)
+			const dataParsers = (typeof dataParser === "function" ? [dataParser] : dataParser).filter(d => d)
+
+			// Make sure that we have arrays
+			if (!Array.isArray(dataUrls) || !Array.isArray(dataKeys) || !Array.isArray(dataParsers))
+				return
+
+			// Only fetch the data endpoints that we have url, key, and parser for
+			const numDataFetchers = Math.min(dataKeys.length, dataUrls.length, dataParsers.length);
+
+			const fetchData = []
+
+			// Generate the promise for fetching data
+			for (let i = 0; i < numDataFetchers; i++) {
+				// Get the url, key, and parser for this entry
+				const dataUrlEntry = dataUrls[i]
+				const dataKeyEntry = dataKeys[i]
+				const dataParserEntry = dataParsers[i]
+
+				// Get the cache key
+				const fetchCacheKey = `${dataKeyEntry}-${dataUrlEntry}-${btoa(dataParserEntry.toString())}`
+
+				// If this data has already been fetched, then simply get it from the cache
+				if (fetchCache[fetchCacheKey]) fetchData.push(Promise.resolve(fetchCache[fetchCacheKey]))
+
+				// Otherwise we load it and save it to the cache
+				else fetchData.push(
+					fetch(dataUrlEntry)
+						.then(response => response.text())
+						.then(text => dataParserEntry(text))
+						.then(data => {
+							setFetchCache({...fetchCache, [fetchCacheKey]: data})
+							return Promise.resolve(data)
+						})
+				)
+			}
+
+			// Wait for all the data to be loaded (or attempted to be loaded)
+			Promise.allSettled(fetchData).then(parsedData => {
+				const data = {}
+
+				// Save each successful response value into the data object
+				parsedData.forEach(({ value }, i) => {
+					if (value) data[dataKeys[i]] = value
+				})
+
+				// Set the data and unset loading
+				setData(data)
+				setLoading(false)
+			})
 		})()
-	}, [dataUrl, dataParser])
+	}, [dataUrl, dataParser, dataKey])
 
 	// If data hasn't loaded yet, return the loading component instead
-	if (loading) return loadingComponent
+	if (loading && loadingComponent) return loadingComponent
```

Comment:
The `loadingComponent` pattern that we were doing here felt a little redundant and less than ideal, but instead of reworking the entire thing I simply made it so that it could be disabled with a falsy attribute. Open to ideas here if folks have thoughts.

Reply:
The reason we have the loading component here is to avoid a page reflow when a visualization loads by rendering a component that is the same visual size as the charts will be once the data loads. If you have a way to restructure that more elegantly, I'd be happy to see a PR for that!